util/sst_file_manager_impl.cc

```diff
@@ -118,15 +150,44 @@ bool SstFileManagerImpl::EnoughRoomForCompaction(
     }
   }
 
-  if (max_allowed_space_ != 0 &&
-      (size_added_by_compaction + cur_compactions_reserved_size_ +
-           total_files_size_ + compaction_buffer_size_ >
-       max_allowed_space_)) {
-    return false;
-  }
   // Update cur_compactions_reserved_size_ so concurrent compaction
   // don't max out space
+  // needed_headroom_ is based on current size reserved by compactions,
+  // minus any files created by running compactions as they would count
+  // against the reserved size.
   cur_compactions_reserved_size_ += size_added_by_compaction;
+  size_t needed_headroom =
+      cur_compactions_reserved_size_ - in_progress_files_size_ +
+      (compaction_buffer_size_ > 0 ? compaction_buffer_size_
+                                   : reserved_disk_buffer_);
+  if (max_allowed_space_ != 0 &&
+      (needed_headroom + total_files_size_ > max_allowed_space_)) {
+    cur_compactions_reserved_size_ -= size_added_by_compaction;
+    return false;
+  }
```

Comment:
This is a fairly complicated logic. I wonder whether we can simplify it, trading off some feature supports.

Reply:
There are a couple of things I added here that increased the complexity - 1. The in_progress_files_size_ was added in order to account for a compaction that has generated some outputs that are already counted in cur_compactions_reserved_size_, and 2. Since compaction_buffer_size_ is a user provided parameter during ```SstFileManager``` construction, choose a default if they didn't set it. 

I can move these down below to the ```NoSpace``` case in order to avoid impacting the normal case. Maybe #2 can also be removed, depending on how conservative we want to be when estimating headroom after getting out of space error.